### PR TITLE
v0.9 Fix build failure due to rust-lang/rust PR 99413

### PR DESCRIPTION
**Pull request by steffahn**
v0.9 branch: Fixes build failure on current `beta` (1.64).

### CODE REVIEW

**Comment by steffahn** [comment]
It would make sense to review this together with #2966 and #2967.

I’m unfamiliar with this code, and since there’s some `unsafe` blocks involved, someone needs to confirm that the change in drop order isn’t a soundness issue.

These PRs modify the `wgpu-core` crate, which would need new point releases 0.7.2, 0.8.2, and 0.9.3.

File: wgpu-core/src/device/mod.rs [hunk 1]
```diff
@@ -1483,6 +1483,7 @@ impl<B: GfxBackend> Device<B> {
 
         // `BTreeMap` has ordered bindings as keys, which allows us to coalesce
         // the descriptor writes into a single transaction.
+        let mut desc_set; // early declaration so it's dropped before write_map
```
Comment (cwfitzgerald) [inline wgpu-core/src/device/mod.rs:1486]: This comment is inverted - it gets dropped after write_map when unwinding.

What was the error in the first place?

**Review by cwfitzgerald — CHANGES_REQUESTED** [review]
Normally I wouldn't do patches for these old crates, but a build failure seems a good enough reason, so once we get these sorted I'll patch them.

File: wgpu-core/src/device/mod.rs (hunk 1 above)
Comment (steffahn) [reply wgpu-core/src/device/mod.rs:1486]: Eh, right, the comment is the wrong way around 😅 

Build error is
```
error[E0597]: `desc_set` does not live long enough
    --> wgpu-core/src/device/mod.rs:1792:26
     |
1792 |                     set: desc_set.raw_mut(),
     |                          ^^^^^^^^^^^^^^^^^^ borrowed value does not live long enough
...
1813 |     }
     |     -
     |     |
     |     `desc_set` dropped here while still borrowed
     |     borrow might be used here, when `write_map` is dropped and runs the `Drop` code for type `BTreeMap`
     |
     = note: values in a scope are dropped in the opposite order they are defined
```

The reason is the PR mentioned in the title, a soundness fix for `BTreeMap` which previously behaved (to dropck) as if no `K` and `V` values would be dropped when a `BTreeMap<K, B>` is dropped.

**Comment by steffahn** [comment]
I fixed the typo/inversion of the comment in the PRs :-)

**Comment by steffahn** [comment]
FYI, I found this breakage in this crater run report: https://crater-reports.s3.amazonaws.com/pr-99413/index.html

_(click on “regressed: dependencies (29 root crates, 322 regressed crates in total)”, then search for “wgpu-core”)_

**Review by cwfitzgerald — APPROVED** [review]
Alright, this looks like it was a legit soundness issue before, and shouldn't have any major issues unless I'm totally mis-remembering how gfx-hal works. So I'll merge these and get these out later today. Thanks for the proactiveness!

**Comment by cwfitzgerald** [comment]
@jimblandy why are there extra PRs on here?

**Comment by jimblandy** [comment]
I added them to get it to build with current Rust.

**Comment by jimblandy** [comment]
The remaining CI problems are test runs, not build failures. Since this is an old release, I think we should merge the PR anyway.

**Comment by m4b** [comment]
@cwfitzgerald can we get a crates release for this when you get the chance? i've tested locally and compiles fine for me now on 1.64 + 1.65, thanks!

**Pull request merged**